{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import gtfparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_cpat_best_translations(best_orfs, orf_fasta):\n",
    "    best_orf_ids = best_orfs['ID'].unique()\n",
    "    best_orf_sequences = []\n",
    "    best_orf_map = {}\n",
    "    \n",
    "    for record in SeqIO.parse(orf_fasta, 'fasta'):\n",
    "        if record.id in best_orf_ids:\n",
    "            transcript = record.id.split('|')[0]\n",
    "            record.id=transcript\n",
    "            record.description=''\n",
    "            record.name=''\n",
    "            record.seq = record.seq.translate(stop_symbol='')\n",
    "            best_orf_sequences.append(record)\n",
    "\n",
    "            best_orf_map[transcript] = record\n",
    "  \n",
    "    with open('GENCODE.CPAT.best.translations.fa', 'w') as ofile:\n",
    "        SeqIO.write(best_orf_sequences, ofile, 'fasta')\n",
    "    return best_orf_map\n",
    "\n",
    "cpat_best_orfs = pd.read_table('./GENCODE.ORF_prob.best.tsv')\n",
    "orf_fasta = './GENCODE.ORF_seqs.fa'\n",
    "gencode_cpat_best_map  = get_cpat_best_translations(cpat_best_orfs, orf_fasta)\n",
    "gencode_sequences = {}\n",
    "for record in SeqIO.parse('./gencode.v35.pc_translations.fa', 'fasta'):\n",
    "    transcript = record.id.split('|')[1]\n",
    "    gencode_sequences[transcript] = record\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_reference_to_cpat(reference, cpat):\n",
    "    sequence_comparisons = []\n",
    "    for key in reference.keys():\n",
    "        if key in cpat.keys():\n",
    "            is_match = reference[key].seq == cpat[key].seq\n",
    "            sequence_comparisons.append([key, is_match, str(reference[key].seq), str(cpat[key].seq)])\n",
    "    else:\n",
    "        sequence_comparisons.append([key, False, str(reference[key].seq), ''])\n",
    "    return sequence_comparisons\n",
    "\n",
    "sequence_comparisons = compare_reference_to_cpat(gencode_sequences, gencode_cpat_best_map)\n",
    "sequence_comparisons_df = pd.DataFrame(sequence_comparisons, columns=['Transcript', 'Match', 'Ref-Seq','CPAT-Seq'])\n",
    "sequence_comparisons_df.groupby('Match').size()\n",
    "sequence_comparisons_df.to_csv('./GENCODE.CPAT.COMPARISIONS.tsv', sep='\\t', index=False)\n",
    "\n",
    "\n",
    "gtf_filename = './gencode.v35.basic.annotation.gtf'\n",
    "gtf = gtfparse.read_gtf(gtf_filename)\n",
    "\n",
    "\n",
    "transcripts = gtf[gtf['feature'] == 'transcript']\n",
    "transcripts = transcripts[[\n",
    "    'seqname', 'source', 'strand', 'gene_type', \n",
    "    'gene_name', 'transcript_id', 'transcript_type', \n",
    "    'transcript_name','transcript_support_level', \n",
    "    'tag', 'protein_id']]\n",
    "sequence_comparison= sequence_comparisons_df.merge(transcripts, how = 'left', left_on='Transcript', right_on='transcript_id')\n",
    "\n",
    "\n",
    "\n",
    "def first_n_match(sequence, other_sequence, n):\n",
    "    if type(sequence) is not str or type(other_sequence) is not str:\n",
    "        print(sequence)\n",
    "        print(other_sequence)\n",
    "        return False\n",
    "    if len(sequence) >= n and len(other_sequence) >= n:\n",
    "        return sequence[:n] == other_sequence[:n]\n",
    "    elif len(sequence) == len(other_sequence):\n",
    "        n = len(sequence)\n",
    "        return sequence[:n] == other_sequence[:n]\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def last_n_match(sequence, other_sequence, n):\n",
    "    if type(sequence) is not str or type(other_sequence) is not str:\n",
    "        print(sequence)\n",
    "        print(other_sequence)\n",
    "        return False\n",
    "    if len(sequence) >= n and len(other_sequence) >= n:\n",
    "        return sequence[-n:] == other_sequence[-n:]\n",
    "    elif len(sequence) == len(other_sequence):\n",
    "        n = len(sequence)\n",
    "        return sequence[-n:] == other_sequence[-n:]\n",
    "    else:\n",
    "        return False\n",
    " \n",
    "\n",
    "\n",
    "sequence_comparison = pd.read_table('./GENCODE.CPAT.comparison_with_meta.tsv')\n",
    "sequence_comparison.dropna(subset=['CPAT-Seq'], inplace=True)\n",
    "for n in [5,10,25]:\n",
    "    sequence_comparison[f'first_{n}_match'] = sequence_comparison.apply(lambda row: first_n_match(row['Ref-Seq'], row['CPAT-Seq'], n), axis = 1)\n",
    "    sequence_comparison[f'last_{n}_match'] = sequence_comparison.apply(lambda row: last_n_match(row['Ref-Seq'], row['CPAT-Seq'], n), axis = 1)\n",
    "sequence_comparison.to_csv('GENCODE.CPAT.comparison_with_meta.tsv', sep='\\t', index=False)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "2f40891b4156b07d6f7078c8cafe72a6146e3c2223ae724d19ed7912498cb5b1"
  },
  "kernelspec": {
   "display_name": "Python 3.8.8 64-bit ('lrpg': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
